contracts/oracles/Oracle.sol

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import {IUniswapV2Factory} from "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import {IUniswapV2Pair} from "@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol";
+import {UniswapV2OracleLibrary} from "./external-modules/uniswap-v2/v2-periphery/UniswapV2OracleLibrary.sol";
+
+import {ArrayHelper} from "../libs/arrays/ArrayHelper.sol";
+
+/**
+ * @title PriceFeedOracle
+ * @dev A contract for retrieving price feeds from Uniswap V2 pairs.
+ */
+abstract contract Oracle is Initializable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using UniswapV2OracleLibrary for address;
+    using ArrayHelper for uint256[];
+    using Math for uint256;
+
+    struct PairInfo {
+        uint256[] prices0Cumulative;
+        uint256[] prices1Cumulative;
+        uint256[] blockTimestamps;
+    }
+
+    IUniswapV2Factory public uniswapV2Factory;
+
+    uint256 public timeWindow;
+
+    EnumerableSet.AddressSet internal _pairs;
+    mapping(address => uint256) internal _pairCounters;
```

Comment:
Maybe we can store this counter in a `PairInfo` structure?

Reply:
Yep, great idea, thank you)